src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,148 @@
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node<T> currentNext = search(index);
+        Node<T> currentPrevious = currentNext.previous;
+        Node<T> newTail = new Node<>(currentPrevious, value, currentNext);
+        if (currentPrevious != null) {
+            currentPrevious.next = newTail;
+        }
+        currentNext.previous = newTail;
+        if (index == 0) {
+            head = newTail;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return search(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = search(index);
+        T currentOld = current.value;
+        current.value = value;
+        return currentOld;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = search(index);
+        Node<T> currentNext = current.next;
+        Node<T> currentPrevious = current.previous;
+        unlink(currentPrevious, currentNext, index);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value == object || object != null && object.equals(current.value)) {
+                Node<T> currentNext = current.next;
+                Node<T> currentPrevious = current.previous;
+                unlink(currentPrevious, currentNext, i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size && size != 0 || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Houston we have a BUG!");
+        }
+    }
+
+    private Node<T> search(int index) {
+        checkIndex(index);
+        Node<T> current;
+        if (size / 2 > index) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.previous;
+            }
+        }
+        return current;
+    }
```

Comment:
try to split this method into two different methods! (searchFromHead, searchFromTail)